### PR TITLE
Incremental kapt

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 1]
```diff
@@ -87,13 +88,24 @@ class NoOpFactoryAnnotationProcessor : AbstractProcessor() {
...
+                packageName,
+                "$className.kt",
+                typeElement
+            )
+        val writer = kotlinFileObject.openWriter()
+        file.writeTo(writer)
+        writer.close()
 
-        processingEnv.messager.printMessage(Diagnostic.Kind.NOTE, "Generated ${file.name}")
+        processingEnv.messager.printMessage(Diagnostic.Kind.WARNING, "Generated ${file.name}")
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:108]: why WARNING ?

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 2]
```diff
@@ -87,13 +88,24 @@ class NoOpFactoryAnnotationProcessor : AbstractProcessor() {
...
 
-        val file = FileSpec.builder(packageName, typeSpec.name.orEmpty())
+        val className = typeSpec.name.orEmpty()
+        val file = FileSpec.builder(packageName, className)
             .addType(typeSpec)
+            .indent("    ")
             .build()
 
-        file.writeTo(File(kaptKotlinGeneratedDir))
+        val kotlinFileObject = processingEnv.filer
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:97]: nice....where did you get that ?

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt [hunk 3]
```diff
@@ -87,13 +88,24 @@ class NoOpFactoryAnnotationProcessor : AbstractProcessor() {
...
+            .createResource(
+                StandardLocation.SOURCE_OUTPUT,
+                packageName,
+                "$className.kt",
+                typeElement
+            )
+        val writer = kotlinFileObject.openWriter()
```
Comment (mariusc83) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:104]: I am wondering if we should use a try/finally here to make sure this is close ?

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt (hunk 2 above)
Comment (xgouchet) [reply tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:97]: in the annotation processing documentation

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt (hunk 3 above)
Comment (xgouchet) [reply tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:104]: In theory those writer methods should be safe.

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt (hunk 1 above)
Comment (xgouchet) [reply tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactoryAnnotationProcessor.kt:108]: It was for debug purpose

**Pull request merged**